lib/plugins/package/lib/zipService.js

```diff
@@ -1,73 +1,188 @@
+ * @param {String[]} patterns Array of globs to use for file matching
+ */
+function getFilesToZip(servicePath, patterns) {
+  return BbPromise.fromCallback((cb) => {
+    // sync version returns temp zip files from .serverless dir
+    glob(patterns, {
+      cwd: servicePath,
+      dot: true,
+      silent: true,
+      follow: true,
+    }, cb);
+  }).filter((filePath) => {
+    const stats = fs.statSync(path.resolve(servicePath, filePath));
+    return !stats.isDirectory(filePath);
+  });
+}
 
-    const output = fs.createWriteStream(artifactFilePath);
+/**
+ * Checks whether machine has zip binary in PATH
+ * return {Promise.<Boolean>} Is available - true/false
+ */
+function isZipAvailable() {
+  return new BbPromise((resolve) => {
+    which('zip', (error) => resolve(!error));
+  });
+}
 
-    output.on('open', () => {
-      zip.pipe(output);
+/**
+ * Get a list of arguments for zip command each less than ~100k chars
+ * Format: ["'folder/file1' 'file2' 'file3'", ...]
+ *
+ * Because of MAX_ARG_STRLEN which causes 'spawn E2BIG' error in node.js
+ * we chunk one huge command string into smaller
+ * @param {String[]} filePaths Array of file paths to zip
+ * @return {String[]} Array of arguments for zip
+ */
+function getZipArgsParts(filePaths) {
+  const MAX_ARG_STRLEN = 128 * 1024;
+  // Safety measure: don't push to the limits
+  const maxArgLength = MAX_ARG_STRLEN - (16 * 1024);
 
-      const files = glob.sync(patterns, {
-        cwd: servicePath,
-        dot: true,
-        silent: true,
-        follow: true,
-      });
+  return filePaths.reduce((memo, filePath) => {
+    const partIndex = memo.length - 1;
+    const willFitIntoArgsPart = memo[partIndex].length < maxArgLength;
 
-      files.forEach((filePath) => {
-        const fullPath = path.resolve(
-          servicePath,
-          filePath
-        );
+    if (willFitIntoArgsPart) {
+      memo[partIndex] += `'${filePath}' `; // eslint-disable-line no-param-reassign
+    } else {
+      memo.push(`'${filePath}' `);
+    }
 
-        const stats = fs.statSync(fullPath);
+    return memo;
+  }, ['']);
+}
 
-        if (!stats.isDirectory(fullPath)) {
-          zip.append(fs.readFileSync(fullPath), {
-            name: filePath,
-            mode: stats.mode,
-          });
+/**
+ * Zips given file paths into the destination path
+ *
+ * zip command supports globs by -i / -x args but it's not used deliberately
+ * as it doesn't support ! negations, overrides and
+ * compromises compatibility with the fallback which relies on files list from glob-all
+ * @param {String} artifactFilePath Destination path of zip
+ * @param {String[]} filePaths Array of file paths to zip
+ * @return {Promise.<String>} Destination path of zip
+ */
+function zipWithNative(artifactFilePath, filePaths) {
+  const filesArgParts = getZipArgsParts(filePaths);
+
+  // Important: .each ensures it's running serially
+  return BbPromise.each(filesArgParts, (filesArgPart) => {
```

Comment:
Not quite -- I may have commented at the wrong part of the code, but in a project with 6 different artifacts (one for each function) I get 6 zip processes running in parallel, even though my computer only has 4 cores. I think that the number of artifacts being zipped at a time should be limited to the number of cores.